### PR TITLE
Data structures and utilities for (rolling horizon) dynamic optimization

### PR DESCRIPTION
**Pull request by Robbybp**
## Summary/Motivation:
I've had some data structures and utilities for dynamic optimization and rolling horizon case studies sitting on my hard drive for a while, and thought it was finally time to clean them up a little, write tests, and commit them somewhere. See the Overview doc page for my motivation for putting together this contrib package. I've found this code useful for MPC case studies, open-loop simulation case studies, and for formulating different kinds of optimal control problems.

## Changes proposed in this PR:
- Add `mpc` contrib package with data structures, utilities, and an interface class for dynamic optimization
- Add tests and some documentation

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/mpc/interfaces/load_data.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    time_indices = [
+        find_nearest_index(time_list, t, tolerance=tolerance)
+        for t in data.get_time_points()
+    ]
+    for idx, t in zip(time_indices, data.get_time_points()):
+        if idx is None:
+            raise RuntimeError("Time point %s not found time set" % t)
+    if len(time_list) != len(data.get_time_points()):
+        raise RuntimeError(
+            "TimeSeriesData object and model must must have same number"
```
Comment (blnicho) [inline pyomo/contrib/mpc/interfaces/load_data.py:66]: Typo: "must must"

File: pyomo/contrib/mpc/interfaces/model_interface.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+
+        """
+        return {
+            cuid: var.value
+            for cuid, var in zip(self._scalar_var_cuids, self._scalar_vars)
+        }
+
+    def get_data_at_time(self, time=None, include_expr=False):
+        """
+        Gets data at a single time point or set of time point. Note that
```
Comment (blnicho) [inline pyomo/contrib/mpc/interfaces/model_interface.py:126]: Typo: "set of time **points**"

File: pyomo/contrib/mpc/modeling/input_constraints.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def get_piecewise_constant_constraints(
+        inputs,
+        time,
+        sample_points,
+        use_next=True,
+        ):
+    """
+    """
```
Comment (blnicho) [inline pyomo/contrib/mpc/modeling/input_constraints.py:23]: Missing docstring

File: pyomo/contrib/mpc/modeling/terminal.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+def get_penalty_expressions_at_time(
+    variables,
+    time,
+    t,
+    target_data,
+    weight_data=None,
+):
+    """
+    """
```
Comment (blnicho) [inline pyomo/contrib/mpc/modeling/terminal.py:46]: Missing docstring

**Review by blnicho — APPROVED** [review]
I found a few typos but otherwise this looks fine.

**Comment by Robbybp** [comment]
After discussing with @kuanhanl, including on PRs https://github.com/Robbybp/pyomo/pull/5 and https://github.com/kuanhanl/pyomo/pull/1, I've made some changes that I think will make this framework more amenable to MHE. These include:
- Updating the functions in cost_expressions.py and terminal.py to return a set indexing the list of provided variables in addition to the expression, which is now indexed by this set as well as time
- Updating names of "tracking cost" functions to refer to a generic "penalty" rather than tracking cost

In addition, I've made some API changes that make things somewhat simpler:
- Remove `load_data_from_...` methods on `DynamicInterface` class. Now there is just a single `load_data` function that automatically decides how to load the data based on its type
- Add a `get_penalty_from_target` function and method, which constructs an appropriate quadratic penalty expression depending on the type of data provided for the target

File: doc/OnlineDocs/contributed_packages/mpc/overview.rst [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+The key should uniquely correspond to a component, or "sub-component"
+that is indexed only by time. A slice, e.g. ``m.var[:, "A"]`` seems
+natural. However, Pyomo provides a better data structure that can
+be constructed from a component, slice, or string, called
+``ComponentUID``. Being constructable from a string is important as
+we may want to store or serialize this data in a form that is agnostic
+of any particular ``ConcreteModel`` object.
+We can now generate our data structure as:
+
+.. code:: python
```
[the remark this answers is not in the document]
Comment (Robbybp) [reply doc/OnlineDocs/contributed_packages/mpc/overview.rst:86]: When I wrote this, I was a little uncertain about how doctests work. I didn't realize that they would behave as a continuation of the previous python "script"/doctest, so I thought this code would be invalid as a standalone snippit. Have updated to make this a doctest.

File: pyomo/contrib/mpc/modeling/terminal.py (hunk 4 above)
Comment (Robbybp) [reply pyomo/contrib/mpc/modeling/terminal.py:47]: I've converted this function to be private and added a small description in the docstring.

**Comment by Robbybp** [comment]
@blnicho Thank you for looking at this thoroughly. I've addressed your comments, and this is ready for re-review.

**Comment by Robbybp** [comment]
OSX failure appears unrelated

File: pyomo/contrib/mpc/interfaces/tests/test_interface.py [hunk 6]
```diff
@@ -189,7 +195,9 @@ def test_load_data_from_dict_indexed_var_list_data(self):
         interface = DynamicModelInterface(m, m.time)
         data_list = [2, 3, 4]
         data = {pyo.ComponentUID(m.input): data_list}
-        interface.load_data(data)
+        # Need to provided data to load_data that can be interpreted
```
Comment (blnicho) [inline pyomo/contrib/mpc/interfaces/tests/test_interface.py:198]: Typo: provide

File: pyomo/contrib/mpc/interfaces/tests/test_interface.py [hunk 7]
```diff
@@ -389,24 +464,199 @@ def test_get_tracking_cost_from_constant_setpoint_var_subset(self):
             m.input[:]: 0.01,
         })
 
-        m.tracking_cost = interface.get_tracking_cost_from_constant_setpoint(
+        variables = [m.var[:, "A"], m.var[:, "B"]]
+        m.variable_set = pyo.Set(initialize=range(len(variables)))
+        #new_set, tr_cost = interface.get_penalty_from_constant_target(
```
Comment (blnicho) [inline pyomo/contrib/mpc/interfaces/tests/test_interface.py:469]: Can this commented line be removed?

File: pyomo/contrib/mpc/interfaces/tests/test_interface.py (hunk 6 above)
Comment (Robbybp) [reply pyomo/contrib/mpc/interfaces/tests/test_interface.py]: Thanks, fixed.

**Pull request merged**